Networking/NetworkingTests/Remote/WCAnalyticsCustomerRemoteTests.swift

```diff
@@ -0,0 +1,112 @@
+    }
+
+    override func tearDown() {
+        network = nil
+        remote = nil
+        super.tearDown()
+    }
+
+    func test_WCAnalyticsCustomerRemote_when_calls_retrieveCustomers_then_returns_result_isSuccess() throws {
+        // Given
+        network.simulateResponse(requestUrlSuffix: "customers", filename: "wc-analytics-customers")
+
+        // When
+        let result = waitFor { promise in
+            self.remote.retrieveCustomers(for: self.sampleSiteID) { result in
+                promise(result)
+            }
+        }
+
+        // Then
+        XCTAssert(result.isSuccess)
+    }
+
+    func test_WCAnalyticsCustomerRemote_when_calls_retrieveCustomers_fails_then_returns_result_isFailure() {
+        // Given
+        network.simulateError(requestUrlSuffix: "", error: NetworkError.notFound)
+
+        // When
+        let result = waitFor { promise in
+            self.remote.retrieveCustomers(for: self.sampleSiteID) { result in
+                promise(result)
+            }
+        }
+
+        // Then
+        XCTAssert(result.isFailure)
+    }
+
+    func test_WCAnalyticsCustomerRemote_when_calls_retrieveCustomerByName_then_returns_result_isSuccess() throws {
+        // Given
+        network.simulateResponse(requestUrlSuffix: "customers?search=\(sampleCustomerName)", filename: "wc-analytics-customers")
+
+        // When
+        let result = waitFor { promise in
+            self.remote.retrieveCustomerByName(for: self.sampleSiteID, with: self.sampleCustomerName) { result in
+                promise(result)
+            }
+        }
+
+        // Then
+        XCTAssert(result.isSuccess)
+        assertEqual(true, result.isSuccess)
+    }
+
+    func test_WCAnalyticsCustomerRemote_when_calls_retrieveCustomerByName_fails_then_returns_result_isFailure() {
+        // Given
+        network.simulateError(requestUrlSuffix: "", error: NetworkError.notFound)
+
+        // When
+        let result = waitFor { promise in
+            self.remote.retrieveCustomerByName(for: self.sampleSiteID, with: self.sampleCustomerName) { result in
+                promise(result)
+            }
+        }
+
+        // Then
+        XCTAssert(result.isFailure)
+    }
+
+    func test_WCAnalyticsCustomerRemote_when_calls_retrieveCustomers_then_returns_all_parsed_customers_successfully() throws {
+        // Given
+        network.simulateResponse(requestUrlSuffix: "customers", filename: "wc-analytics-customers")
+
+        // When
+        let result = waitFor { promise in
+            self.remote.retrieveCustomers(for: self.sampleSiteID) { result in
+                promise(result)
+            }
+        }
+        let customers = try XCTUnwrap(result.get())
+
+        // Then
+        XCTAssertNotNil(customers)
+        XCTAssertEqual(customers[0].userID, 1)
+        XCTAssertEqual(customers[1].userID, 2)
+        XCTAssertEqual(customers[0].name, "John")
+        XCTAssertEqual(customers[1].name, "Paul")
+    }
```

Comment:
It's probably worth doing this test for customer search too. Once you add that, you can delete both the `isSuccess` tests, as they're just simpler versions of the same. 

Checking that the customers are actually returned, not just a success result (which may be empty) is valuable, as returning the mapped data is part of the Remote's responsibilities.

You don't necessarily need to check both properties here, as that is tested on the Mapper. Maybe just that the names are as you expect, or that there are two customers in the returned data, but still more than `isSuccess`.

Reply:
Thanks!

> Checking that the customers are actually returned, not just a success result (which may be empty) is valuable, as returning the mapped data is part of the Remote's responsibilities.

Following your advice here and checking what this actually returns, this has uncovered a problem with the way `retrieveCustomerByName()` worked, and that's was always retuning the full array of entities, not the filtered-by-name entities. 

I'm assume this is because our Mapper class just calls `map()`, not other methods from the subclass implementation details. In order to solve this, I moved the filtering logic to the Remote instead: https://github.com/woocommerce/woocommerce-ios/pull/7780/commits/def5158743fcae7d44aeef8c931e23a788d77860 . I'm not positive this is a good way to do it (specially the usage of `.contains()` and `.filter()` ), but seems to resonate better with other cases I've checked in our codebase. Another alternative could have been to create a new mapper entirely for this, but seemed a bit overkill. Happy to get any feedback.

Reply:
Forgot to add: At this point I can't be sure yet how it will work, but my guess is that once all customers are fetched, we could also use [FetchedResults](https://developer.apple.com/documentation/swiftui/fetchedresults) in the Storage layer to request a particular set of results from local storage, rather than having to perform further requests through the network.

Reply:
> I'm assuming this is because our Mapper class just calls map(), not other methods from the subclass implementation details. In order to solve this, I moved the filtering logic to the Remote instead: https://github.com/woocommerce/woocommerce-ios/commit/def5158743fcae7d44aeef8c931e23a788d77860 . I'm not positive this is a good way to do it (specially the usage of .contains() and .filter() ), but seems to resonate better with other cases I've checked in our codebase. Another alternative could have been to create a new mapper entirely for this, but seemed a bit overkill.

Gut feeling without having looked at any of the code: the Networking layer should probably just return what the API does, it shouldn't go beyond correctly mapping entities and making requests and such. The filtering sounds more like business logic.

If the API doesn't give us filtered results back, that business logic should be handled in Yosemite, and as you suggest in your other comment, this may well mean leveraging CoreData's filtering to do it.

Reply:
> Gut feeling without having looked at any of the code: the Networking layer should probably just return what the API does, it shouldn't go beyond correctly mapping entities and making requests and such. The filtering sounds more like business logic.
> 
> If the API doesn't give us filtered results back, that business logic should be handled in Yosemite, and as you suggest in your other comment, this may well mean leveraging CoreData's filtering to do it.

Thanks, that makes a lot of sense. Most of my doubts regarding this comes from trying to understand how Networking and Core Data will work together, but I'm getting a clearer idea as we iterate 👍